File: typescript/examples/quant-tier/webhook-lifecycle/webhook_lifecycle.ts
import { fileURLToPath } from "node:url";
import { parseArgs } from "node:util";

import { AccountWebhooksApi } from "cs2cap";

import { formatApiError } from "../../_shared/api.js";
import { buildConfiguration, loadApiKey, requireRealApiKey } from "../../_shared/auth.js";
import { CleanupRegistry } from "../../_shared/cleanup.js";
import { printHelpIfRequested } from "../../_shared/cli.js";
import { renderTable } from "../../_shared/render.js";

const SCRIPT_PATH = fileURLToPath(import.meta.url);

function readOptions(): { url: string; label: string; keepResources: boolean } {
  const { values } = parseArgs({
    options: {
      url: { type: "string" },
      label: { type: "string" },
      "keep-resources": { type: "boolean" },
    },
  });
  return {
    url: values.url ?? "https://example.com/cs2cap-webhook",
    label: values.label ?? "SDK example webhook",
    keepResources: values["keep-resources"] ?? false,
  };
}

async function main(): Promise<number> {
  if (printHelpIfRequested("Usage: npm run example -- examples/quant-tier/webhook-lifecycle/webhook_lifecycle.ts [--url https://example.com/cs2cap-webhook] [--label \"SDK example webhook\"] [--keep-resources]")) {
    return 0;
  }
  const args = readOptions();
  const cleanup = new CleanupRegistry(args.keepResources);

  let bearerToken: string;
  try {
    bearerToken = loadApiKey("CS2C_API_KEY", SCRIPT_PATH);
    requireRealApiKey(bearerToken);
  } catch (error) {
    console.error(error instanceof Error ? error.message : String(error));
    return 1;
  }

  const configuration = buildConfiguration(bearerToken);

  try {
    const webhooksApi = new AccountWebhooksApi(configuration);
    const before = await webhooksApi.listWebhooks();
    const created = await webhooksApi.createWebhook({
      webhookCreateRequest: { label: args.label, url: args.url, isActive: false },
    });
    const webhookId = created.webhook.id;
    cleanup.add("delete temporary webhook", () => webhooksApi.deleteWebhook({ webhookId }));
    const updated = await webhooksApi.updateWebhook({
      webhookId,
      webhookUpdateRequest: { label: `${args.label} (updated)`, isActive: false },
    });
    const rotated = await webhooksApi.rotateWebhookSecret({ webhookId });
    const deliveries = await webhooksApi.listWebhookDeliveries({ limit: 5 });
    const deliveryDetail = deliveries.deliveries[0]
      ? await webhooksApi.getWebhookDelivery({ deliveryId: deliveries.deliveries[0].id })
      : null;
    const after = await webhooksApi.listWebhooks();
    const cleanupResults = await cleanup.run();

    console.log("Webhook lifecycle (/v1/account/webhooks):");
    console.log(
      renderTable(
        ["step", "webhook_id", "label", "active", "secret_last4"],
        [
          ["before_count", before.webhooks.length, "N/A", "N/A", "N/A"],
          ["created", created.webhook.id, created.webhook.label, String(created.webhook.isActive), created.webhook.secretLast4],
          ["updated", updated.id, updated.label, String(updated.isActive), updated.secretLast4],
          ["rotated", rotated.webhook.id, rotated.webhook.label, String(rotated.webhook.isActive), rotated.webhook.secretLast4],
          ["after_count", after.webhooks.length, "N/A", "N/A", "N/A"],
        ],
      ),
    );
    console.log();
    console.log("Webhook deliveries (/v1/account/webhooks/deliveries):");
    console.log(`- listed: ${deliveries.deliveries.length}`);
    console.log(`- detail_checked: ${deliveryDetail?.id ?? "N/A"}`);
    console.log();
    console.log("Cleanup summary:");
    console.log(
      renderTable(
        ["resource", "status", "detail"],
        cleanupResults.map((result) => [result.label, result.ok ? "ok" : "failed", result.detail]),
      ),
    );
    return 0;
  } catch (error) {
    console.error(await formatApiError(error));
    const cleanupResults = await cleanup.run();
    if (cleanupResults.length > 0) {
      console.error(renderTable(["resource", "status", "detail"], cleanupResults.map((result) => [result.label, result.ok ? "ok" : "failed", result.detail])));
    }
    return 1;
  }
}

void main().then((code) => {
  process.exitCode = code;
});

File: python/examples/quant-tier/webhook-lifecycle/webhook_lifecycle.py
import argparse
import sys
from pathlib import Path

import cs2cap
from cs2cap.rest import ApiException

EXAMPLES_ROOT = Path(__file__).resolve().parents[2]
if str(EXAMPLES_ROOT) not in sys.path:
    sys.path.insert(0, str(EXAMPLES_ROOT))

from _shared.auth import build_configuration, load_api_key, require_real_api_key  # noqa: E402
from _shared.cleanup import CleanupRegistry  # noqa: E402
from _shared.render import render_table  # noqa: E402


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="Create, update, rotate, inspect, and delete a webhook.")
    parser.add_argument("--url", default="https://example.com/cs2cap-webhook", help="Destination URL.")
    parser.add_argument("--label", default="SDK example webhook", help="Temporary webhook label.")
    parser.add_argument("--keep-resources", action="store_true", help="Leave the created webhook active.")
    return parser


def main() -> int:
    args = build_parser().parse_args()
    cleanup = CleanupRegistry(keep_resources=args.keep_resources)

    try:
        bearer_token = load_api_key("CS2C_API_KEY", start=Path(__file__))
        require_real_api_key(bearer_token)
    except RuntimeError as exc:
        print(str(exc), file=sys.stderr)
        return 1

    configuration = build_configuration(bearer_token)

    try:
        with cs2cap.ApiClient(configuration) as client:
            webhooks_api = cs2cap.AccountWebhooksApi(client)
            before = webhooks_api.list_webhooks()
            created = webhooks_api.create_webhook(
                cs2cap.WebhookCreateRequest(label=args.label, url=args.url, is_active=False)
            )
            webhook_id = created.webhook.id
            cleanup.add("delete temporary webhook", lambda: webhooks_api.delete_webhook(webhook_id))
            updated = webhooks_api.update_webhook(
                webhook_id,
                cs2cap.WebhookUpdateRequest(label=f"{args.label} (updated)", is_active=False),
            )
            rotated = webhooks_api.rotate_webhook_secret(webhook_id)
            deliveries = webhooks_api.list_webhook_deliveries(limit=5)
            delivery_detail = (
                webhooks_api.get_webhook_delivery(deliveries.deliveries[0].id)
                if deliveries.deliveries
                else None
            )
            after = webhooks_api.list_webhooks()

    except ApiException as exc:
        print(f"API request failed: {exc}", file=sys.stderr)
        return 1
    finally:
        cleanup_results = cleanup.run()

    print("Webhook lifecycle (/v1/account/webhooks):")
    print(
        render_table(
            ["step", "webhook_id", "label", "active", "secret_last4"],
            [
                ["before_count", str(len(before.webhooks)), "N/A", "N/A", "N/A"],
                ["created", created.webhook.id, created.webhook.label, str(created.webhook.is_active), created.webhook.secret_last4],
                ["updated", updated.id, updated.label, str(updated.is_active), updated.secret_last4],
                ["rotated", rotated.webhook.id, rotated.webhook.label, str(rotated.webhook.is_active), rotated.webhook.secret_last4],
                ["after_count", str(len(after.webhooks)), "N/A", "N/A", "N/A"],
            ],
        )
    )
    print()
    print("Webhook deliveries (/v1/account/webhooks/deliveries):")
    print(f"- listed: {len(deliveries.deliveries)}")
    print(f"- detail_checked: {delivery_detail.id if delivery_detail else 'N/A'}")
    print()
    print("Cleanup summary:")
    print(render_table(["resource", "status", "detail"], [[r.label, "ok" if r.ok else "failed", r.detail] for r in cleanup_results]))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
